README
Miredo : open-source Teredo
============================
Copyright (C) 2004-2007 Rémi Denis-Courmont.

  Welcome to Miredo’s user’s guide!

Where to go?
=============

  If you’ve used olders versions of this program, you should read NEWS
for a summary of the most recent updates and changes.

  For detailled usage instructions, you should refer to the Unix manual
page miredo(8) which should be provided with your copy of the program.
For quick usage instructions, see below.

  See INSTALL for general instructions on how to build the package and
install the program from sources. Additionnal informations may be found
below. If you are building from the Subversion repository, run the
“./autogen.sh” script first.

  This package is distributed under the terms of the General Public
License (GPL) version 2 written by the Free Software Foundation, Inc.
for full licensing details, please read COPYING.

  If you have further questions, please send an email to:
    miredo (dash) devel (at) remlab (dot) net


What is Miredo?
================

  Miredo is an Unix daemon program which mostly implements the
“Teredo: Tunneling IPv6 over UDP through NATs” Internet proposed
standard (RFC 4380). It can provide either client or relay
functionality. A separate program, miredo-server is also included in
the package; it consists of a Teredo server.

  Miredo can be used to provide IPv6 connectivity to users behind NAT
devices, such as broadband routers. Most of these device do not support
IPv6, and do not allow forwarding of proto-41 (including 6to4).


System requirements
====================

  Miredo aims to support all POSIX-like open-source operating system
kernels with IPv6 and userland layer-3 tunneling support. See below
for system specific notes.

  Miredo is believed to be architecture-independant, but it was only
properly tested on Linux i386, with GCC 4.1, GCC 3.3 and ICC 8.0.
GCC 2.95 is purposedly not supported.

  When available, Miredo can use the following optional libraries :
 - GNU gettext for localization,
 - libcap (currently Linux-specific) for POSIX capabilities,
 - Judy dynamic arrays library for better scalability.

Linux:
-------
  Miredo runs fine on Linux kernel 2.6.9; if possible, kernel version
2.6.12 or more recent is recommended. It can run with older versions as
well, including the 2.4 branch. However, there are known problems with
the -experimental- IPv6 stack of these kernel releases; this is not
specific to the usage of Miredo.
  Miredo requires the Universal TUNTAP driver (CONFIG_TUN) and IPv6
(CONFIG_IPV6).

  For proper source address selection (RFC3484 support), kernel version
2.6.17.2 and GNU/libc 2.5 are required.

FreeBSD:
---------
  Miredo works fine with FreeBSD 5.5 and up. You can use the net/miredo
port. There is a known problem with the FreeBSD 6.2 and older; it
should mostly affect Miredo in "relay" mode.
  FreeBSD 4.11 is not supported.

OpenBSD:
---------
  OpenBSD might work. Version 3.7 was tested with older releases.

NetBSD:
--------
  NetBSD 4.0 is required, for IPv6 tunneling. Miredo should be
available through pkgsrc.

Darwin (Mac OS X):
-------------------
  Miredo works on Mac OS X (Panther, Tiger) and might work on other
Darwin variants. On Mac OS X, you will have to install a third party
tunneling driver; Miredo was tested with Mattis Nissler's tuntap from:
http://www-user.rhrk.uni-kl.de/~nissler/tuntap/

  Note that Mac OS X users should consider using Miredo-OSX from
http://www.deepdarc.com/ which provides prebuilt OS X binaries of
Miredo along with Nissler's tuntap, LibJudy and OS X specific additions
and tweaks.

DragonFly:
-----------
  Miredo should work. It is in NetBSD's pkgsrc.


Quick usage
============

Easy installation:
-------------------
  First, compile and install Miredo. Refer to INSTALL for detailled
instructions. Miredo can be installed the usual way:

# ./configure
# make
# su
# make install

  Miredo has no particular required dependencies, besides the usual
C/C++ compilers and development libraries.

  A sample configuration file is automatically installed at
/usr/local/etc/miredo.conf - unless the file already existed (which
means you are probably reinstalling or upgrading Miredo). This sample
will cause Miredo to run as a Teredo client, with “teredo.remlab.net”
(Miredo official testing Teredo server) as its Teredo server. These
default settings should be fine for most users.

Starting the program:
----------------------
  Before you start, please note that Miredo must be started by root,
and that it will detach and run in the background. If something goes
wrong, there are two ways two know what :
- read your system logs (typically /var/log/syslog),
- force Miredo to run in the foreground (that’s meant for debugging),
  by starting it with the “--foreground” command line parameter, and
  wait for about 20 seconds.

  You can now run miredo (as root!):
# /usr/local/sbin/miredo

  It will need some time to initialize, particularly if you are behind
a restricted NAT, which is frequent. After about 20 seconds, you should
have access to the IPv6 Internet through Teredo, with a public Teredo
IPv6 address on the “teredo” networking interface :

# ifconfig teredo
teredo    Link encap:UNSPEC  HWaddr 00-00-00-00-00-00-00-00-00-00-00...
          inet6 addr: 2001:0:8ac3:9ddd:0:7ffa:ad80:3464/32 Scope:...
          inet6 addr: fe80::5445:5245:444f/64 Scope:Link
          UP POINTOPOINT RUNNING NOARP  MTU:1280  Metric:1
          RX packets:5 errors:0 dropped:0 overruns:0 frame:0
          TX packets:7 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:500
          RX bytes:468 (468.0 b)  TX bytes:560 (560.0 b)

# ping6 -c 4 www.kame.net
PING www.kame.net(orange.kame.net) 56 data bytes
64 bytes from orange.kame.net: icmp_seq=1 ttl=50 time=558 ms
64 bytes from orange.kame.net: icmp_seq=2 ttl=50 time=585 ms
64 bytes from orange.kame.net: icmp_seq=3 ttl=50 time=562 ms
64 bytes from orange.kame.net: icmp_seq=4 ttl=50 time=552 ms

--- www.kame.net ping statistics ---
4 packets transmitted, 4 received, 0% packet loss, time 3003ms
rtt min/avg/max/mdev = 552.830/564.865/585.031/12.218 ms

Monitoring:
------------
  If you wish to monitor the Teredo tunnel, I suggest you use famous
network analyzer Wireshark (formerly Ethereal®) which includes a Teredo
“dissector”.

Teredo relay and/or server:
----------------------------
  Please refer to the sample configuration miredo.conf-dist for further
information. You can get a comprehensive reference of all possible
options in the manual pages provided with the package:
miredo(8) and miredo.conf(5)

# man 8 miredo
# man 5 miredo.conf


Securing you Miredo installation
=================================

  By default, Miredo drops its root privileges and runs as user nobody.
While that is far more secure than keeping root privileges as previous
versions did by default, it is not optimal. If you are security
concious, paranoid, or if you are building a package, you are advised
to perform the following steps to restrict the impact of a possible
compromise of the Miredo daemon.

  They are some steps to secure Miredo installation, because they are
non trivial and non portable, they cannot be done automatically. That
is why miredo defaults to using “nobody” user account which is
available on any POSIX-like operating system.

1) System user:
----------------
  Miredo should run with its own user account rather than common user
“nobody”. They are two ways to do that :

- You can enable the “--enable-miredo-user” command line option when
  running the source code configure script. If you are a packager,
  please use that method. Miredo will try to SetUID as “miredo” by
  default, though that can be overriden with the “-u” command line
  option (see man page miredo(8)).

- You can use the “-u” option when starting Miredo. That saves the
  cost of recompiling Miredo. For example:

  # /usr/local/sbin/miredo -u miredo

NOTE: If you are running Miredo as a Teredo client, Miredo will spawn a
separate privileged process whose only job will be the Teredo interface
parameters (it must be root to do that). If someone breaks Miredo, it
might still be able to break your IPv6 networking setup, but it should
not be able to compromise the whole system.

2) Chroot jail:
----------------
  Chroot jail can be enabled with the “-t” command line option. Note
that when using miredo as a Teredo client, you will typically have to
install your DNS resolver library and configuration within the chroot
environment, which is why the feature is currently disabled by default.
  On Linux/libc6, that would consist of copying /etc/resolv.conf,
/etc/nsswitch.conf and the Name Service Switch shared objects within
the chroot. You will also have to resynchronize /etc/resolv.conf within
the chroot with the one at the system root every time it is modified
(such as when the DHCP client updates /etc/resolv.conf).

  If you intend to use miredo only as a Teredo relay and/or server,
you should really enable the chroot, as it is safer and should work
fine “out of the box”.

3) POSIX capabilities:
-----------------------
  Miredo supports POSIX.1e capabilities (at least on Linux), if they
are available. You should not need to worry as it is entirely
automatic.

  If you are a packager, you should consider installing your system’s
POSIX capabilities library development files, before building Miredo.


Feedback:
==========

  If you have further questions, please write to:

    miredo (dash) devel (at) remlab (dot) net

-- 
Rémi Denis-Courmont <remi (at) remlab (dot) net>
http://www.remlab.net/miredo/


NEWS
RELEASE NOTES for Miredo
stable release version 1.2.3
=============================

  This is a stable release. There are no known outstanding issues. We
strongly encourage users from previous releases to upgrade, as some important
fixes are included.
  Please report any problem by email to miredo-devel (at) remlab (dot) net

  You can find major changes since version 0.1.0 below. For more general
informations on this program, see README. For syntax details and
changes, run “miredo -h” to get help. For detailled internal changes, see
ChangeLog.

===========================================================================
STABLE RELEASE 1.2.3 : Minor portability fixes

# Fix C pointer aliasing violations.
# Fix compilation with FreeBSD 7.2 or later.

===========================================================================
STABLE RELEASE 1.2.2 : Major bug fixes

# Fix SIGHUP signal handling (again).
# Compilation fix for Debian GNU/kFreeBSD.

===========================================================================
STABLE RELEASE 1.2.1 : Minor bug fixes

# Fix deadlock if Miredo fails to start.

===========================================================================
STABLE RELEASE 1.2.0 : Minor portability fix

# Fix compilation on FreeBSD

===========================================================================
STABLE RELEASE 1.1.7 : Major bug fixes

# Fix crash in client more (regression from 1.1.6).
# Thread-safety fixes (with regards to sigaction and strerror).
# Fix client hook script for FreeBSD

===========================================================================
STABLE RELEASE 1.1.6 : Minor bug fixes

# Fix compiling on FreeBSD.
# Incomplete ISATAP support removed.
  In the mean time, proper support was added within the Linux kernel.
# Fix client error handling when resolving server name.
# Fix server link-local address, compatibility with Vista clients.

===========================================================================
STABLE RELEASE 1.1.5 : Minor bug fixes

# Fix SIGHUP signal handling.
# Fix compiling on MacOS X and/or without libJudy.

===========================================================================
STABLE RELEASE 1.1.4 : Minor bug fixes

# Fix client-hook script problems with Fedora.
# Fix unlikely race condition in Teredo maintenance procedure setup.

===========================================================================
STABLE RELEASE 1.1.3 : Major bug fixes, minor feature enhancement

# Fix Teredo address randomization (reachability problems).
# Better debug messages.
# Allow GNU General Public License version 3 (and would be later).

===========================================================================
TEST RELEASE 1.1.2 : Minor bug fixes

# Fix filtering error within miredo-server (regression from 1.1.0).

===========================================================================
TEST RELEASE 1.1.1 : Minor bug fixes

# Fix fatal padding mistake on old ABI ARM achitecture.

===========================================================================
TEST RELEASE 1.1.0 : Major features enhancement

# Remove NAT type determination:
  Miredo now runs behind any type of NAT. However, connectivity might
  be severely degraded behind the worst devices, such as symmetric NATs.

# Add 12 bits of randomness to Teredo client address:
  Teredo addresses are less predictible, which should enhance host
  protection against network scanning.

# Remove brittle and battery-unfriendly "autoclient" mode:
  Proper default IPv6 source address selection (RFC3484) implementation
  would address most of the use cases for this, as well as other issues.
  If that is not sufficient, an external connection management system
  is anyway needed to start/stop Miredo when appropriate.

# Restore "cone" RelayType:
  All relays should use it, as it improves support for some kinds of
  (pretty broken but nevertheless deployed) NAT devices.

# Minor specification conformance fixes to miredo-server.
# Use a hook shell script for client interface configuration.
# Add a bunch of debug messages to debug builds.
# Rewrite clock subsystem to avoid polling when idle (battery savings).
# Some minor bug fixes.

===========================================================================
STABLE RELEASE 1.0.6 : Minor features enhancement

# Fix support for Teredo peers behind symmetric NATs (bug from 0.9.8).
  This could probably be used to spoof a Teredo clients.
# MacOS X compilation fix.

===========================================================================
STABLE RELEASE 1.0.5 : Major bug fixes

# Fix issues with timer and compiler optimizations (bug from 0.9.4).
# Increase ping test hop limit:
  Some IPv6 native nodes could not be reached otherwise.
# Increase HMAC secret size from 64 to 128 bits.
# Fix filtering of packets with link-local source address (bug from 0.4.0).
# Various fixes to the experimental ISATAP daemon.

============================================================================
STABLE RELEASE 1.0.4 : Major bug fix

# Fix packet storm with Teredo client behind symmetric NATs.

============================================================================
STABLE RELEASE 1.0.3 : Minor feature enhancement

# Work-around for compatibility with Microsoft Teredo servers.
# Better detection of symmetric NATs and access network changes.
# Limit spamming of syslog.

============================================================================
STABLE RELEASE 1.0.2 : Minor portability and bug fixes

# Several portability and minor/impossible bug fixes (see ChangeLog).

============================================================================
STABLE RELEASE 1.0.1 : Minor portability fixes

# Install configuration file samples into a dedicated directory.
# Build fixes (libteredo would not link on Mac OS X).

============================================================================
STABLE RELEASE 1.0.0 : Minor portability fixes


src/relay.c
/*
 * relay.c - Miredo: binding between libtun6 and libteredo
 * $Id$
 */

/***********************************************************************
 *  Copyright © 2004-2006 Rémi Denis-Courmont.                         *
 *  This program is free software; you can redistribute and/or modify  *
 *  it under the terms of the GNU General Public License as published  *
 *  by the Free Software Foundation; version 2 of the license.         *
 *                                                                     *
 *  This program is distributed in the hope that it will be useful,    *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of     *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.               *
 *  See the GNU General Public License for more details.               *
 *                                                                     *
 *  You should have received a copy of the GNU General Public License  *
 *  along with this program; if not, you can get it from:              *
 *  http://www.gnu.org/copyleft/gpl.html                               *
 ***********************************************************************/

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <assert.h>
#include <gettext.h>

#include <inttypes.h>
#include <stdlib.h> // free()
#include <stdio.h> // fputs()
#include <sys/types.h>
#include <string.h> // strerror()
#include <errno.h>
#include <unistd.h> // close()
#include <fcntl.h>
#include <sys/wait.h> // wait()
#include <sys/select.h> // pselect()
#include <signal.h> // sigemptyset()
#include <compat/pselect.h>
#include <syslog.h>
#include <pthread.h>

#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/ip6.h>
#include <netinet/icmp6.h>
#include <arpa/inet.h> // inet_ntop()
#include <netdb.h> // NI_MAXHOST
#ifdef HAVE_SYS_CAPABILITY_H
# include <sys/capability.h>
#endif
#ifndef SOL_IPV6
# define SOL_IPV6 IPPROTO_IPV6
#endif
#ifndef SOL_ICMPV6
# define SOL_ICMPV6 IPPROTO_ICMPV6
#endif

#include <libtun6/tun6.h>

#include <libteredo/teredo.h>
#include <libteredo/tunnel.h>

#include "privproc.h"
#include "addrwatch.h"
#include "miredo.h"
#include "conf.h"


static int relay_diagnose (void)
{
	char errbuf[LIBTUN6_ERRBUF_SIZE];
	if (tun6_driver_diagnose (errbuf))
	{
		fputs (errbuf, stderr);
		return -1;
	}

	return 0;
}


typedef struct miredo_tunnel
{
	tun6 *tunnel;
	int priv_fd;
	teredo_tunnel *relay;
} miredo_tunnel;

static int icmp6_fd = -1;

static int miredo_init (bool client)
{
	if (teredo_startup (client))
		return -1;

	assert (icmp6_fd == -1);

	icmp6_fd = socket (AF_INET6, SOCK_RAW, IPPROTO_ICMPV6);
	if (icmp6_fd == -1)
		return -1;

	miredo_setup_nonblock_fd (icmp6_fd);

	setsockopt (icmp6_fd, SOL_IPV6, IPV6_CHECKSUM, &(int){2}, sizeof (int));

	/* We don't use the socket for receive -> block all */
	struct icmp6_filter filt;
	ICMP6_FILTER_SETBLOCKALL (&filt);
	setsockopt (icmp6_fd, SOL_ICMPV6, ICMP6_FILTER, &filt, sizeof (filt));
	return 0;
}


static void miredo_deinit (bool client)
{
	assert (icmp6_fd != -1);
	close (icmp6_fd);
	teredo_cleanup (client);
}


/**
 * Callback to transmit decapsulated Teredo IPv6 packets to the kernel.
 */
static void
miredo_recv_callback (void *data, const void *packet, size_t length)
{
	assert (data != NULL);

	(void)tun6_send (((miredo_tunnel *)data)->tunnel, packet, length);
}


/**
 * Callback to emit an ICMPv6 error message through a raw ICMPv6 socket.
 */
static void
miredo_icmp6_callback (void *data, const void *packet, size_t length,
                        const struct in6_addr *dst)
{
	(void)data;
	assert (icmp6_fd != -1);

	struct sockaddr_in6 addr =
	{
		.sin6_family = AF_INET6,
#ifdef HAVE_SA_LEN
		.sin6_len = sizeof (struct sockaddr_in6),
#endif
		//.sin6_addr = *dst TODO: make sure dst is aligned
	};
	memcpy (&addr.sin6_addr, dst, sizeof (addr.sin6_addr));

	/* TODO: use sendmsg and don't memcpy in BuildICMPv6Error */
	(void)sendto (icmp6_fd, packet, length, 0,
	              (struct sockaddr *)&addr, sizeof (addr));
}


//#define TEREDO_CONE     0
#define TEREDO_RESTRICT 1
#define TEREDO_CLIENT   2
#define TEREDO_EXCLIENT 3

static bool
ParseRelayType (miredo_conf *conf, const char *name, int *type)
{
	unsigned line;
	char *val = miredo_conf_get (conf, name, &line);

	if (val == NULL)
		return true;

	if (strcasecmp (val, "client") == 0)
		*type = TEREDO_CLIENT;
	else
	if (strcasecmp (val, "autoclient") == 0)
		*type = TEREDO_EXCLIENT;
	else
	if ((strcasecmp (val, "relay") == 0)
	 || (strcasecmp (val, "cone") == 0)
	 || (strcasecmp (val, "restricted") == 0))
		*type = TEREDO_RESTRICT;
	else
	{
		syslog (LOG_ERR, _("Invalid relay type \"%s\" at line %u"),
		        val, line);
		free (val);
		return false;
	}
	free (val);
	return true;
}


#ifdef MIREDO_TEREDO_CLIENT
static tun6 *
create_dynamic_tunnel (const char *ifname, int *fd)
{
	tun6 *tunnel = tun6_create (ifname);
	if (tunnel == NULL)
		return NULL;

	/* FIXME: we leak all heap-allocated settings in the child process */
	int res = miredo_privileged_process (tunnel);
	if (res == -1)
	{
		tun6_destroy (tunnel);
		return NULL;
	}
	*fd = res;
	return tunnel;
}


static void
destroy_dynamic_tunnel (tun6 *tunnel, int fd)
{
	assert (fd != -1);
	close (fd);

	wait (NULL); // wait for privsep process

	tun6_destroy (tunnel);
}


/**
 * Callback to configure a Teredo tunneling interface.
 */
static void
miredo_up_callback (void *data, const struct in6_addr *addr, uint16_t mtu)
{
	char str[INET6_ADDRSTRLEN];

	syslog (LOG_NOTICE, _("Teredo pseudo-tunnel started"));
	if (inet_ntop (AF_INET6, addr, str, sizeof (str)) != NULL)
		syslog (LOG_INFO, _(" (address: %s, MTU: %u)"),
		        str, (unsigned)mtu);

	assert (data != NULL);

	miredo_configure_tunnel (((miredo_tunnel *)data)->priv_fd, addr, mtu);
}


/**
 * Callback to deconfigure a Teredo tunneling interface.
 */
static void
miredo_down_callback (void *data)
{
	assert (data != NULL);

	miredo_configure_tunnel (((miredo_tunnel *)data)->priv_fd, &in6addr_any,
	                         1280);
	syslog (LOG_NOTICE, _("Teredo pseudo-tunnel stopped"));
}


static int
setup_client (teredo_tunnel *client, const char *server, const char *server2)
{
	teredo_set_state_cb (client, miredo_up_callback, miredo_down_callback);
	return teredo_set_client_mode (client, server, server2);
}
#else
# define create_dynamic_tunnel( a, b )   NULL
# define destroy_dynamic_tunnel( a, b )   (void)0
# define setup_client( a, b, c )         (-1)
# define miredo_addrwatch_available( a ) 0
# define miredo_addrwatch_getfd( a )     (-1)
# define run_tunnel( a, b )           run_tunnel_RELAY_ONLY( a )
#endif


static tun6 *
create_static_tunnel (const char *restrict ifname,
                      const struct in6_addr *restrict prefix,
                      uint16_t mtu)
{
	tun6 *tunnel = tun6_create (ifname);

	if ((tunnel == NULL) && (ifname != NULL) && (errno == ENOSYS))
		tunnel = tun6_create (NULL);
	if (tunnel == NULL)
		return NULL;

	if (tun6_setMTU (tunnel, mtu) || tun6_bringUp (tunnel)
	 || tun6_addAddress (tunnel, &teredo_restrict, 64)
	 || tun6_addRoute (tunnel, prefix, 32, 0))
	{
		tun6_destroy (tunnel);
		return NULL;
	}
	return tunnel;
}


#if 0 && !defined (MIREDO_DEFAULT_USERNAME)
static void
destroy_static_tunnel (tun6 *restrict tunnel,
                       const struct in6_addr *restrict prefix)
{
	/*
	 * Manual clean up of the tunnel device is only possible if we retain root
	 * privileges. Unfortunately, if we don't do this, it becomes impossible
	 * to restart Miredo on FreeBSD. The proper fix belongs in the FreeBSD
	 * kernel - that's the only way to handle kills (or would-be crashes) of
	 * Miredo; that's part of the standard job of a solid kernel process
	 * killer.
	 */
	tun6_delRoute (tunnel, prefix, 32, 0);
	tun6_delAddress (tunnel, &teredo_restrict, 64);
	tun6_bringDown (tunnel);
	tun6_destroy (tunnel);
}
#else
# define destroy_static_tunnel( t, p ) tun6_destroy( t )
#endif


static int
setup_relay (teredo_tunnel *relay, uint32_t prefix)
{
	teredo_set_prefix (relay, prefix);
	return teredo_set_relay_mode (relay);
}


/**
 * Thread to encapsulate IPv6 packets into UDP.
 * Cancellation safe.
 */
static LIBTEREDO_NORETURN void *miredo_encap_thread (void *d)
{
	teredo_tunnel *relay = ((miredo_tunnel *)d)->relay;
	tun6 *tunnel = ((miredo_tunnel *)d)->tunnel;

	for (;;)
	{
		/* Handle incoming data */
		union
		{
			struct ip6_hdr ip6;
			uint8_t fill[65507];
		} pbuf;

		/* Forwards IPv6 packet to Teredo
		 * (Packet transmission) */
		int val = tun6_wait_recv (tunnel, &pbuf.ip6, sizeof (pbuf));
		if (val >= 40)
		{
			pthread_setcancelstate (PTHREAD_CANCEL_DISABLE, NULL);
			teredo_transmit (relay, &pbuf.ip6, val);
			pthread_setcancelstate (PTHREAD_CANCEL_ENABLE, NULL);
		}
		else
			pthread_testcancel ();
	}
}


/**
 * Miredo main daemon function, with UDP datagrams and IPv6 packets
 * receive loop.
 */
static int
run_tunnel (miredo_tunnel *tunnel, miredo_addrwatch *w)
{
	int fd = miredo_addrwatch_getfd (w);
	if (fd >= (int)FD_SETSIZE)
		return -1;

	pthread_t encap_th;
	if (teredo_run_async (tunnel->relay)
	 || pthread_create (&encap_th, NULL, miredo_encap_thread, tunnel))
		return -1;

	/* Main loop */
	int retval = -2;

	while (!miredo_addrwatch_available (w))
	{
		fd_set readset;
		FD_ZERO (&readset);
		if (fd != -1)
			FD_SET (fd, &readset);

		sigset_t set;
		sigemptyset (&set);
		if (pselect (fd + 1, &readset, NULL, NULL, NULL, &set) < 0)
		{
			retval = 0;
			break;
		}
	}

	pthread_cancel (encap_th);
	pthread_join (encap_th, NULL);
	return retval;
}


static int
relay_run (miredo_conf *conf, const char *server_name)
{
	/*
	 * CONFIGURATION
	 */
	union teredo_addr prefix;
	memset (&prefix, 0, sizeof (prefix));
	prefix.teredo.prefix = htonl (TEREDO_PREFIX);

	int mode = TEREDO_CLIENT;
	if (!ParseRelayType (conf, "RelayType", &mode))
	{
		syslog (LOG_ALERT, _("Fatal configuration error"));
		return -2;
	}

#ifdef MIREDO_TEREDO_CLIENT
	const char *server_name2 = NULL;
	char namebuf[NI_MAXHOST], namebuf2[NI_MAXHOST];
#endif
	uint16_t mtu = 1280;

	if (mode & TEREDO_CLIENT)
	{
#ifdef MIREDO_TEREDO_CLIENT
		if (server_name == NULL)
		{
			char *name = miredo_conf_get (conf, "ServerAddress", NULL);
			if (name == NULL)
			{
				syslog (LOG_ALERT, _("Server address not specified"));
				syslog (LOG_ALERT, _("Fatal configuration error"));
				return -2;
			}
			strlcpy (namebuf, name, sizeof (namebuf));
			free (name);
			server_name = namebuf;

			name = miredo_conf_get (conf, "ServerAddress2", NULL);
			if (name != NULL)
			{
				strlcpy (namebuf2, name, sizeof (namebuf2));
				free (name);
				server_name2 = namebuf2;
			}
		}
#else
		syslog (LOG_ALERT, _("Unsupported Teredo client mode"));
		syslog (LOG_ALERT, _("Fatal configuration error"));
		return -2;
#endif
	}
	else
	{
		server_name = NULL;
		mtu = 1280;

		if (!miredo_conf_parse_teredo_prefix (conf, "Prefix",
		                                      &prefix.teredo.prefix)
		 || !miredo_conf_get_int16 (conf, "InterfaceMTU", &mtu, NULL))
		{
			syslog (LOG_ALERT, _("Fatal configuration error"));
			return -2;
		}
	}

	uint32_t bind_ip = INADDR_ANY;
	uint16_t bind_port = 
#if 0
		/*
		 * We use 3545 as a Teredo service port.
		 * It is better to use a fixed port number for the
		 * purpose of firewalling, rather than a pseudo-random
		 * one (all the more as it might be a "dangerous"
		 * often firewalled port, such as 1214 as it happened
		 * to me once).
		 */
		IPPORT_TEREDO + 1;
#else
		0;
#endif

	if (!miredo_conf_parse_IPv4 (conf, "BindAddress", &bind_ip)
	 || !miredo_conf_get_int16 (conf, "BindPort", &bind_port, NULL))
	{
		syslog (LOG_ALERT, _("Fatal configuration error"));
		return -2;
	}

	bind_port = htons (bind_port);

	char *ifname = miredo_conf_get (conf, "InterfaceName", NULL);

	miredo_conf_clear (conf, 5);

	/*
	 * SETUP
	 */

	// Tunneling interface initialization
	int privfd = -1;
	tun6 *tunnel = (mode & TEREDO_CLIENT)
		? create_dynamic_tunnel (ifname, &privfd)
		: create_static_tunnel (ifname, &prefix.ip6, mtu);

	if (ifname != NULL)
		free (ifname);

	int retval = -1;

	if (tunnel == NULL)
	{
		syslog (LOG_ALERT, _("Miredo setup failure: %s"),
		        _("Cannot create IPv6 tunnel"));
		return -1;
	}

	if (miredo_init ((mode & TEREDO_CLIENT) != 0))
		syslog (LOG_ALERT, _("Miredo setup failure: %s"),
		        _("libteredo cannot be initialized"));
	else
	{
#ifdef MIREDO_TEREDO_CLIENT
		miredo_addrwatch *watch = (mode == TEREDO_EXCLIENT)
			? miredo_addrwatch_start (tun6_getId (tunnel)) : NULL;
#endif
		if (drop_privileges () == 0)
		{
			do
			{
				if (miredo_addrwatch_available (watch))
				{
					sigset_t sig;
					sigemptyset (&sig);

					int fd = miredo_addrwatch_getfd (watch);
					assert (fd != -1);

					fd_set rdset;
					FD_ZERO (&rdset);
					FD_SET (fd, &rdset);

					if (pselect (fd + 1, &rdset, NULL, NULL, NULL, &sig) == 1)
						retval = -2;
					else
						retval = 0;

					continue;
				}

				teredo_tunnel *relay = teredo_create (bind_ip, bind_port);
				if (relay != NULL)
				{
					miredo_tunnel data = { tunnel, privfd, relay };
					teredo_set_privdata (relay, &data);
					teredo_set_recv_callback (relay, miredo_recv_callback);
					teredo_set_icmpv6_callback (relay, miredo_icmp6_callback);

					retval = (mode & TEREDO_CLIENT)
						? setup_client (relay, server_name, server_name2)
						: setup_relay (relay, prefix.teredo.prefix);
	
					/*
					 * RUN
					 */
					if (retval == 0)
					{
						retval = run_tunnel (&data, watch);
#ifdef MIREDO_TEREDO_CLIENT
						if (retval == -2)
						{
							assert (mode == TEREDO_EXCLIENT);
							miredo_down_callback (&data);
						}
#endif
					}
					teredo_destroy (relay);
				}
			}
			while (retval == -2);

			if (retval)
				syslog (LOG_ALERT, _("Miredo setup failure: %s"),
				        _("libteredo cannot be initialized"));
		}

#ifdef MIREDO_TEREDO_CLIENT
		if (watch != NULL)
			miredo_addrwatch_stop (watch);
#endif
		miredo_deinit ((mode & TEREDO_CLIENT) != 0);
	}

	if (mode & TEREDO_CLIENT)
		destroy_dynamic_tunnel (tunnel, privfd);
	else
		destroy_static_tunnel (tunnel, &prefix.ip6);

	return retval;
}


extern void miredo_setup_fd (int fd)
{
	(void) fcntl (fd, F_SETFD, FD_CLOEXEC);
}


extern void miredo_setup_nonblock_fd (int fd)
{
	int flags = fcntl (fd, F_GETFL);
	if (flags == -1)
		flags = 0;
	(void) fcntl (fd, F_SETFL, O_NONBLOCK | flags);
	miredo_setup_fd (fd);
}


int main (int argc, char *argv[])
{
#ifdef HAVE_LIBCAP
	static const cap_value_t capv[] =
	{
		CAP_NET_ADMIN, /* required by libtun6 */
		CAP_NET_RAW /* required for raw ICMPv6 socket */
	};

	miredo_capv = capv;
	miredo_capc = sizeof (capv) / sizeof (capv[0]);
#endif

	miredo_name = "miredo";
	miredo_diagnose = relay_diagnose;
	miredo_run = relay_run;

	return miredo_main (argc, argv);
}

